## 1. Objects and Classes

- Create a seeded random number generator with `np.random.default_rng()`
- Define a class `Experiment` that stores a `name` and a NumPy array of `values`, with a method `summary()` that returns the mean and standard deviation of `values`
- Create two `Experiment` objects with simulated data from `rng.normal()`, each with a different `loc` and `scale`, and call `summary()` on each
- In a Markdown cell, answer: which parts of `Experiment` are attributes and which are methods? How close is each `summary()` to the `loc` and `scale` you used?

In [2]:
import matplotlib.pyplot as plt
import numpy as np

In [12]:
rng = np.random.default_rng(seed = 42)

class Experiment:
    def __init__(self, name, values):
        self.name = name
        self.values = np.array(values)
    def summary(self):
        mean = np.mean(self.values)
        std = np.std(self.values)
        return mean, std
        
# from module 3: 
#  - `rng.normal(loc, scale, size)`: draw from a Normal distribution
#  - `loc`: the mean
#  - `scale`: the standard deviation
#  - `size`: the shape of the output (an int, or a tuple for multiple dimensions)

simulated_data1 = rng.normal(122, 12, size=(1,100))
simulated_data2 = rng.normal(342, 23, size=(1,100))

data_test_1 = Experiment('Data 1', simulated_data1)
data_test_2 = Experiment('Data 2', simulated_data2)

data_test_1.summary(), data_test_2.summary()


((np.float64(121.39676466219369), np.float64(9.273408900069404)),
 (np.float64(341.7554677301183), np.float64(22.41588726668262)))

From Class Notes:
Classes, also called Objects, are containers in Python that can store values, known as `attributes`; and functions, known as `methods`
Attributes are accessed with a `.` and no parentheses. Methods are called with `()`, just like functions.

Based on this, we know that name and values are attributes, while the only method is summary. 
Summary is relatively close to the mean and std defined in the random normal distrubutions I made, the 122 mean and 12 std became 121 and 9.27, which are pretty reasonable, though the std. is slightly lower than expected. Simialrly the other random normal distrubution had a mean of 342 and std of 23, and this was well maintained with summary giving a mean of 341.8 and std of 22.4, which are very reasonable values.  

## 2. Loading Data

- Load `notebooks/data/RNAcountTable_chr1_geneNames.tsv` into a DataFrame called `df` with `pd.read_csv()`. The columns are separated by tabs, and your path depends on where your notebook is saved relative to the `data` folder.
- Show the first 5 and the last 5 rows
- Print the `shape`, `columns`, and `dtypes` of `df`
- In a Markdown cell, answer: how many genes and how many samples are in the data? What does each row represent, and what does each number in the table represent?

In [15]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [16]:
df = pd.read_csv('notebooks/data/RNAcountTable_chr1_geneNames.tsv', sep='\t')

FileNotFoundError: [Errno 2] No such file or directory: 'notebooks/data/RNAcountTable_chr1_geneNames.tsv'